rest-service/manager_rest/rest/resources_v3_1/operations.py

```diff
@@ -94,13 +95,32 @@ def patch(self, operation_id, **kwargs):
             if not instance.is_nop and \
                     old_state not in TERMINATED_STATES and \
                     instance.state in TERMINATED_STATES:
-                execution = self._get_execution(sm, instance)
-                if execution and execution.finished_operations is not None:
-                    execution.finished_operations += 1
-                sm.update(execution, modified_attrs=('finished_operations',))
-            instance = sm.update(instance)
+                self._increase_executions_finished_count(instance)
+            instance = sm.update(instance, modified_attrs=('state',))
         return instance
 
+    def _increase_executions_finished_count(self, operation):
+        """Increase finished_operations for this operation's execution
+
+        This is a separate sql-level update query, rather than ORM-level
+        calls, for performance: the operation state-update call is on
+        the critical path for all operations in a workflow; this saves
+        about 3ms over the ORM approach (which requires fetching the
+        execution; more if the DB is not local).
+        """
+        exc_table = models.Execution.__table__
+        tg_table = models.TasksGraph.__table__
+        db.session.execute(
+            exc_table.update()
+            .where(db.and_(
```

Comment:
note that the original had `execution.finished_operations is not None:` but we're not adding a `is null` check here: this is because null+1=null in sql, so they're effectively ignored anyway

(and, in practice, there aren't any executions who have operations, while not having the counts set to 0)